tests/FSharp.Compiler.ComponentTests/ErrorMessages/TypeMismatchTests.fs

```diff
@@ -7,6 +7,52 @@ open FSharp.Test.Compiler
 
 module ``Type Mismatch`` =
 
+    module ``Different tuple lengths`` =
+        
+        [<Fact>]
+        let ``Known type on the left``() =
+            FSharp """
+let x: int * int * int = 1, ""
+let x: int * string * int = "", 1
+let x: int * int = "", "", 1
+            """
+            |> typecheck
+            |> shouldFail
+            |> withDiagnostics [
+                (Error 1, Line 2, Col 26, Line 2, Col 31,
+                 "Type mismatch. Expecting a\n    'tuple of length 3 (int * int * int)'    \nbut given a\n    'tuple of length 2'    \n")
```

Comment:
Could we have the inferred types of the right side in the error, please? :)

Reply:
Well, I tried to do that initially by re-triggering the type check on the tuple types against the same number of inferrence variables. But we discussed it with @dsyme and he didn't think it's a good idea or worth adding this complexity or potential issues (it worked fine on constant expressions but not sure about more complicated scenarios).

Reply:
Oh, that's unfortunate. The original issue was opened mostly due to we wanted to have the types on the right available.

Reply:
Yeah, I think they will be available in case if tuples are of the same length, but if it differs, it will short circuit, and won't infer types for them.

Reply:
@dsyme Could you tell more about what potential issues you see? I've done a somewhat similar type check with new inference variables for extra patterns in https://github.com/dotnet/fsharp/pull/7711, and it seems it didn't have any significant consequences while it improved the user experience a lot.

Reply:
@auduchinok I think you're possibly suggesting a route where the mismatch in tuple lengths is detected much earlier, in CheckExpressions case for `SynExpr.Tuple`, rather than in the constraint solver, and that at that point we have many more choices, similar to what you've done for patterns???

Reply:
@auduchinok For example, at `SynExpr.Tuple` we could

* check if the expected type is already known to be a tuple
* if there's a mismatch, check the tuple expression anyway, all with inference variables as input (rather than using a prefix of elements of the known type), making sure we use error recovery, and perhaps suppressing diagnostics
* report the error with respect to that extracted type information 

I think that would be ok.  @0101 Perhaps we could pair on Monday and sketch that out

Reply:
So more like the code was in https://github.com/dotnet/fsharp/pull/13347/commits/4078e92d64cb4dab8ba949481c7a0c3dd100427f but detecting the upcoming error condition before submitting to the constraint solver (rather than catching the failure from the constraint solver - which could come from nested types, for example)

Reply:
> @auduchinok https://github.com/dotnet/fsharp/pull/7711 is interesting - I didn't review it much. There were some code comments by @TIHan asking questions about the code - could you take a moment to answer them please - thanks!

Yes, I should've definitely replied there earlier (and had some comments in the first place), as now it'll take much more time to recover the context. 🙂 Thank you for reminding about it, I'll try to look there more carefully next week.

> @auduchinok I think you're possibly suggesting a route where the mismatch in tuple lengths is detected much earlier, in CheckExpressions case for SynExpr.Tuple, rather than in the constraint solver, and that at that point we have many more choices, similar to what you've done for patterns???

To be honest it was about any expression on the right, though I'm not sure how feasible that could be.

Reply:
Thanks for looking into this. I could not figure out good way of error recovery that contains the type as part of the error message.  So looking forward to see what it is the outcome of this and learn how to do it in the future  :)

Reply:
It didn't occur to me to do the checking before the constraint solver because it seemed like duplicating the checking logic. But it may be the way to go. 

@dsyme I probably won't be able to this week, but I'd be up for it next week.